File: algorithms/2_medium/K/00779_k_th_symbol_in_grammar/recursion.c
static int helper(const int n, const int k)
{
    if (n == 1) {
        if (k != 1) {
            fprintf(stderr, "unreachable! invalid k");
            exit(1);
        }
        return 0;
    }
    const int parent_pos = (k + 1) / 2;
    const int parent_val = helper(n - 1, parent_pos);

    return ((k & 1) == 0) ? (1 - parent_val) : (parent_val);
}

int kthGrammar(const int n, const int k)
{
    assert(n > 0 && n <= 30 && k > 0);
    return helper(n, k);
}

File: algorithms/2_medium/K/00779_k_th_symbol_in_grammar/recursion.rs
impl Solution {
    fn helper(n: i32, k: i32) -> i32 {
        if n == 1 {
            if k != 1 {
                unreachable!("invalid k");
            }
            return 0;
        }
        let parent_pos = (k + 1) / 2;
        let parent_val = Self::helper(n - 1, parent_pos);

        if (k & 1) == 0 { 1 - parent_val } else { parent_val }
    }

    pub fn kth_grammar(n: i32, k: i32) -> i32 {
        assert!(n > 0 && n <= 30 && k > 0);
        return Self::helper(n, k);
    }
}
